File: backend/app/schemas/environment.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field


# ============================================================
# CREATE ENVIRONMENT
# ============================================================

class EnvironmentCreate(BaseModel):

    name: str = Field(
        ...,
        min_length=2,
        max_length=100,
    )

    key: str = Field(
        ...,
        min_length=2,
        max_length=100,
    )

    description: str | None = Field(
        default=None,
        max_length=500,
    )

    is_active: bool = True


# ============================================================
# UPDATE ENVIRONMENT
# ============================================================

class EnvironmentUpdate(BaseModel):

    name: str | None = Field(
        default=None,
        min_length=2,
        max_length=100,
    )

    key: str | None = Field(
        default=None,
        min_length=2,
        max_length=100,
    )

    description: str | None = Field(
        default=None,
        max_length=500,
    )

    is_active: bool | None = None


# ============================================================
# ENVIRONMENT RESPONSE
# ============================================================

class EnvironmentResponse(BaseModel):

    id: int
    name: str
    key: str
    description: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# ROLE DETAILS
# ============================================================

class EnvironmentUserRoleResponse(BaseModel):

    id: int
    name: str
    description: str | None = None
    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# USER DETAILS
# ============================================================

class EnvironmentUserResponse(BaseModel):

    id: int
    name: str
    email: EmailStr
    is_active: bool

    role: EnvironmentUserRoleResponse

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# RETRIEVED BY
# ============================================================

class EnvironmentRetrievedByResponse(BaseModel):

    user: EnvironmentUserResponse


# ============================================================
# CREATE RESPONSE
# ============================================================

class EnvironmentCreateResponse(BaseModel):

    success: bool
    message: str
    environment: EnvironmentResponse
    created_by: EnvironmentRetrievedByResponse


# ============================================================
# UPDATE RESPONSE
# ============================================================

class EnvironmentUpdateResponse(BaseModel):

    success: bool
    message: str
    environment: EnvironmentResponse
    updated_by: EnvironmentRetrievedByResponse


# ============================================================
# LIST RESPONSE
# ============================================================

class EnvironmentListResponse(BaseModel):

    success: bool
    message: str
    total: int
    environments: list[EnvironmentResponse]
    retrieved_by: EnvironmentRetrievedByResponse


# ============================================================
# SINGLE RESPONSE
# ============================================================

class EnvironmentSingleResponse(BaseModel):

    success: bool
    message: str
    environment: EnvironmentResponse
    retrieved_by: EnvironmentRetrievedByResponse


# ============================================================
# DELETE RESPONSE
# ============================================================

class EnvironmentDeleteResponse(BaseModel):

    success: bool
    message: str
    environment: EnvironmentResponse
    deleted_by: EnvironmentRetrievedByResponse

File: frontend/src/types/environment.ts

// ============================================================
// ENVIRONMENT TYPES
// ============================================================

// ============================================================
// USER DETAILS
// ============================================================

export interface EnvironmentUser {
  id?: number;
  name?: string;
  email?: string;
  role?: string;
  role_id?: number;
}

// ============================================================
// ENVIRONMENT
// ============================================================

export interface Environment {
  id: number;

  name: string;

  key: string;

  description?: string;

  is_active?: boolean;

  created_by_id?: number;

  created_at?: string;

  updated_at?: string;

  created_by?: EnvironmentUser;

  updated_by?: EnvironmentUser;
}

// ============================================================
// CREATE ENVIRONMENT REQUEST
// ============================================================

export interface EnvironmentCreateRequest {
  name: string;

  key: string;

  description?: string;
}

// ============================================================
// UPDATE ENVIRONMENT REQUEST
// ============================================================

export interface EnvironmentUpdateRequest {
  name?: string;

  key?: string;

  description?: string;

  is_active?: boolean;
}

// ============================================================
// CREATE RESPONSE
// ============================================================

export interface EnvironmentCreateResponse {
  success?: boolean;

  message?: string;

  environment?: Environment;

  created_by?: EnvironmentUser;

  data?:
    | Environment
    | {
        environment?: Environment;
        created_by?: EnvironmentUser;
      };
}

// ============================================================
// UPDATE RESPONSE
// ============================================================

export interface EnvironmentUpdateResponse {
  success?: boolean;

  message?: string;

  environment?: Environment;

  updated_by?: EnvironmentUser;

  data?:
    | Environment
    | {
        environment?: Environment;
        updated_by?: EnvironmentUser;
      };
}

// ============================================================
// DELETE RESPONSE
// ============================================================

export interface EnvironmentDeleteResponse {
  success?: boolean;

  message?: string;

  environment_id?: number;

  deleted_by?: EnvironmentUser;

  data?: {
    environment_id?: number;
    deleted_by?: EnvironmentUser;
  };
}

// ============================================================
// STATUS RESPONSE
// ============================================================

export interface EnvironmentStatusResponse {
  success?: boolean;

  message?: string;

  environment?: Environment;

  updated_by?: EnvironmentUser;

  data?:
    | Environment
    | {
        environment?: Environment;
        updated_by?: EnvironmentUser;
      };
}

// ============================================================
// GET SINGLE ENVIRONMENT RESPONSE
// ============================================================

export interface EnvironmentSingleResponse {
  success?: boolean;

  message?: string;

  environment?: Environment;

  retrieved_by?: EnvironmentUser;

  data?:
    | Environment
    | {
        environment?: Environment;
        retrieved_by?: EnvironmentUser;
      };
}

// ============================================================
// GET ENVIRONMENTS RESPONSE
// ============================================================

export interface EnvironmentListResponse {
  success?: boolean;

  message?: string;

  total?: number;

  environments?: Environment[];

  retrieved_by?: EnvironmentUser;

  data?:
    | Environment[]
    | {
        environments?: Environment[];
        total?: number;
        retrieved_by?: EnvironmentUser;
      };
}

